File: pytictactoe/game.py
import os
from board import Board

class Game:
    # internal use
    state = dict() # empty dictionary
    board = []     # empty list
    players = [-1, 1]
    winner = ''
    # configuration
    player_letters = ['X', 'O']
    board_size = 3
    connect_to_win = 3

    def __init__(self, config:dict = None):
        self.state['activePlayerIdx'] = 0
        self._welcome(config)
        self._configure_game(config)
        self.board = Board(self.board_size)

    def tick(self):
        self._reset_screen()
        self.board.print_board()
        move = self.get_next_move()
        self.update_board(move)
        self.update_player()

    def get_next_move(self):
        move = self._get_user_move_input()
        while not self.is_valid_move(move):
            print('Invalid move, space not available. Try again.')
            move = self._get_user_move_input()
        return move
    
    def is_valid_move(self, possible_move):
        return self.board.state[possible_move[0]][possible_move[1]] == 0

    def parse_move(self, move):
        # must be formatted as 'A1'
        col = ord(move[0]) - ord('A')
        row = int(move[1]) - 1
        # have to verify move is available on the board
        return (row, col)

    def update_board(self, loc):
        self.board.update(loc, self.players[self.state['activePlayerIdx']], self.player_letters[self.state['activePlayerIdx']])

    def update_player(self):
        self.state['activePlayerIdx'] = 1 - self.state['activePlayerIdx']

    def print_winner(self):
        self._reset_screen()
        self.board.print_board()
        print(f'Game Over. {self.winner} wins!')

    def is_over(self):
        board_seg = []
        board = self.board.state
        
        # get each row as a segment
        for row in board:
            board_seg.append(row)
        # transpose and check again to get columns
        for col in zip(*board):
            board_seg.append(col)
        
        # get diagonals - top left to bottom right, moving down
        diag1 = dict()
        for offset in range(0, self.board_size - self.connect_to_win + 1):
            for i in range(0, self.board_size - offset):
                if offset not in diag1:
                    diag1[offset] = []
                diag1[offset].append(board[i+offset][i])

        # get diagonals - top left to bottom right, moving right, start at one to skip main diagonal (is in diag3)
        diag2 = dict()
        for offset in range(1, self.board_size - self.connect_to_win + 1):
            for i in range(0, self.board_size - offset):
                if offset not in diag2:
                    diag2[offset] = []
                diag2[offset].append(board[i][i+offset])

        # get diagonals - bottom left to top right, moving up
        diag3 = dict()
        for offset in range(0, self.board_size - self.connect_to_win + 1):
            for i in range(0, self.board_size - offset):
                if offset not in diag3:
                    diag3[offset] = []
                diag3[offset].append(board[self.board_size - 1 - i - offset][i])

        # get diagonals - bottom left to top right, moving right, start at one to skip main diagonal (is in diag3)
        diag4 = dict()
        for offset in range(1, self.board_size - self.connect_to_win + 1):
            for i in range(0, self.board_size - offset):
                if offset not in diag4:
                    diag4[offset] = []
                diag4[offset].append(board[self.board_size - 1 - i][i+offset])

        for d in diag1.values():
            board_seg.append(d)

        for d in diag2.values():
            board_seg.append(d)

        for d in diag3.values():
            board_seg.append(d)
        
        for d in diag4.values():
            board_seg.append(d)

        for seg in board_seg:
            xcount = 0
            ocount = 0
            for i in range(len(seg)):
                if self.players[0] == seg[i]:
                    xcount = xcount + 1
                else:
                    # reset xcount counter if not equal to ensure win only when N number in a row
                    xcount = 0
                if self.players[1] == seg[i]:
                    ocount = ocount + 1
                else:
                    # reset ocount counter, same as xwin reset
                    ocount = 0
                
                if xcount >= self.connect_to_win or ocount >= self.connect_to_win:
                    self.winner =  self.player_letters[0] if xcount else self.player_letters[1]
                    return True
        return False
    
    # --- Private Methods  ---------------------------------------------------------
    def _reset_screen(self):
        os.system('clear')

    def _welcome(self, config=None):
        if config is not None:
            dbg = config.get('debug',False)
            
        if not dbg:
            self._reset_screen()
            print('==========================')
            print('    Tic Tac Toe 3000!')
            print()
            print('Game Configuration')

    def _configure_game(self, config:dict = None):
        if config is None:
            resp = input('Use custom player letters? y/[n]: ')
            if resp == 'y':
                self._get_player_letters()
            resp = input('What size board? [3]: ')
            if resp == '':
                self.board_size = 3
            else:
                self.board_size = int(resp)
            cont = True
            while cont:
                resp = input(f'How many to connect to win? [{self.board_size}]: ')
                if resp == '':
                    resp = self.board_size
                try:
                    self.connect_to_win = int(resp)
                    cont = False
                except ValueError:
                    continue
        else:
            pass
            self.player_letters = config['player_letters']
            self.board_size = config['board_size']
            self.connect_to_win = config['winning_run_length']

    def _get_player_letters(self):
        player1 = input('Player 1, enter your letter: ')
        player2 = input('Player 2, enter your letter: ')
        self.player_letters = [player1, player2]

    def _is_response_valid(self, new_move_response):
        if len(new_move_response) != 2:
            # must two characters
            return False
        try:
            int(new_move_response[1])
        except ValueError:
            # second character is not an int
            return False
        
        if not new_move_response[0].isalpha():
            # first character is not A-Z
            return False

        if new_move_response[0].islower():
            return False

        return True

    def _get_user_move_input(self):
        resp = input(f"Player {self.player_letters[self.state['activePlayerIdx']]}, enter move: ")
        if resp == 'quit':
            raise SystemExit # completely quit the application
        while not self._is_response_valid(resp):
            print('Invalid input format. Try again. Must be formatted as A1')
            resp = input(f"Player {self.player_letters[self.state['activePlayerIdx']]}, enter move: ")
        return self.parse_move(resp)

File: ctictactoe/game.cpp
#include "game.h"
#include "screencmd.h"
#include <iostream>
#include <string>

using namespace std;

// Default Constructor 
Game::Game(){
    num_players = 2;
    player_letters[0]= 'X';
    player_letters[1] = 'O';
    board_size = 3;
    connect_to_win = 3;
    current_player = 1;
    winner = 0;
}

// Print welcome message to screen
void Game::printWelcome(){
    resetScreen();
    std::cout << "==========================" << std::endl;
    std::cout << "    Tic Tac Toe 3000!" << std::endl;
    std::cout << std::endl;
    std::cout << "Game Configuration" << std::endl;
    return;
}

// Customize cout << operator for this class
ostream & operator << (ostream& out, const Game  &g){
    out << "Game Setup: \n";
    out << "Player 1: " << g.player_letters[0] << "\n" ;
    out << "Player 2: " << g.player_letters[1] << std::endl;
    out << "Board Size: " << g.board_size << std::endl;;
    out << "Run Length to Win: "  << g.connect_to_win << std::endl;
    return out;
}

// Configure game object with user input
void Game::configureGame(){
    cout << "Use custom player letters? y/[n]: ";
    string resp;
    getline(cin, resp);
    if (resp == "y"){
        string buf;
        cout << "\nPlayer 1, enter your letter: ";
        getline(cin, buf);
        player_letters[0] = buf[0];
        cout << "\nPlayer 2, enter your letter: ";
        getline(cin, buf);
        player_letters[1] = buf[0];
    }

    cout << "What board size? [3]: ";
    getline(cin, resp);
    if (resp.empty()){
        board_size = 3;
    }else{
        board_size = stoi(resp);
    }
    cout << "How many to connect to win? [3]: ";
    getline(cin, resp);
    if (resp.empty()){
        connect_to_win = 3;
    }else{
        connect_to_win = stoi(resp);
    }

    return;
}

void Game::getNextMove(int* move_out){
    cout << "Player " << current_player << ", enter a move: ";
    string buf;
    getline(cin, buf);
    parseMove(buf, move_out);
    return;
}

void Game::parseMove(string move_in, int* move_out){
    // input in form A1
    // convert to 0,0
    int row, col;
    col = move_in[0]; // A in A1
    row = move_in[1]; // 1 in A1
    move_out[0] = row - '0' - 1;
    move_out[1] = col - 'A';
    return;
}

void Game::updateBoard(Board *board, int* move){
    board->updateState(move[0], move[1], current_player);
    return;
}

void Game::updateCurrentPlayer(){
    current_player++;
    if (current_player > num_players){
        current_player = 1;
    }
    return;
}

void Game::tick(Board *board){
    resetScreen();
    int players[2] = {1,2};
    int move[2] = {0,0};
    // TODO: change printBoard to not need players array
    board->printBoard(num_players, players, player_letters);
    getNextMove(move);
    updateBoard(board, move);
    updateCurrentPlayer();
    return;
}

bool Game::isOver(Board *board){
    int segment[board_size];
    // Check each row
    for (int i = 0; i < board_size; i++){
        for (int j = 0; j < board_size; j++){
            segment[j] = board->state[board_size*i + j];
        }
        if (segmentHasWinner(segment)){
            return true;
        }
    }
    // check each column
    for (int i = 0; i < board_size; i++){
        for (int j = 0; j < board_size; j++){
            segment[j] = board->state[board_size*j + i];
        }
        if (segmentHasWinner(segment)){
            return true;
        }
    }

    // diagonal 1 - top left to bottom right, moving down
    for (int i = 0; i < board_size; i++){
        segment[i] = board->state[board_size*i + i];
        if (segmentHasWinner(segment)){
            return true;
        }
    }

    // diagonal 2 -  top right to bottom left
    for (int i = 0; i < board_size; i++){
        segment[i] = board->state[board_size-1-i + board_size*i];
        if (segmentHasWinner(segment)){
            return true;
        }
    }

    
    return false;
}

bool Game::segmentHasWinner(int *segment){
    // initialize count var, used to contain number of 
    // player[i] spots in a row on the board
    int count[num_players];
    for (int i = 0; i < num_players; i++){
        count[i] = 0;
    }

    // Loop over segment to look for a winner
    for (int i = 0; i < board_size; i++){
        // If element indicates player1, inc player1 count
        if ((segment[i] > 0) && (segment[i] == 1)){
            count[0]++;
        // If element is no move or is player2, reset count
        }else{
            count[0] = 0;
        }
        // If element indicates player2, inc player2 count
        if ((segment[i] > 0) && (segment[i] == 2)){
            count[1]++;
        // If element is no move or is player1, reset count
        }else{
            count[1] = 0;
        }

        // Check winner each loop
        // winner could be early if connect_to_win < board_size
        if (count[0] >= connect_to_win){
            winner = 1;
            break;
        }
        if (count[1] >= connect_to_win){
            winner = 2;
            break;
        }
    }
    return (winner > 0);
}

void Game::printWinner(){
    resetScreen();
    cout << "Player " << winner << " ( " << player_letters[winner-1] << " ) wins the game!" << endl;
}